### PR TITLE
Add pycodestyle github action to enforce PEP8 rules in the future PR's

### PR DESCRIPTION
**Pull request by technic**
Changes to the codebase were made here
https://github.com/OpenPLi/enigma2/pull/2921

Based on the script here:
https://github.com/OpenPLi/enigma2/pull/2941

Related issue:
https://github.com/OpenPLi/enigma2/issues/2104

Note that due to some reason there are already few rules broken in the current state of the source files. This might be fixed by applying PEP8 script again. 

cc: @persianpros

PS. now only very limited list of linters is selected. There are quite more pep8 rules violated in enigma2 code base (even excluding indents).

### CODE REVIEW

**Comment by persianpros** [comment]
Remember if your result isn't the outcome of PEP8.sh then it's not good and you should close this PR as all enigma2 teams agreed on that so don't change it please.

Also if it's going to do automatic cleanup it will be rejected by PLi.

**Comment by technic** [comment]
> 
> 
> Remember if your result isn't the outcome of PEP8.sh then it's not good and you should close this PR as all enigma2 teams agreed on that so don't change it please.
> 
> Also if it's going to do automatic cleanup it will be rejected by PLi.

This patch checks in the CI (e.g. used before merging a PR), that all new code is conformed to the style given in PEP8.sh script.

**Comment by Huevos** [comment]
This is a crazy idea that will stop even more contributors.

**Comment by technic** [comment]
What is really crazy, is having random code style, no tests, and ancient python2.

**Comment by persianpros** [comment]
Current e2 scene is PEP8 with same style so what's wrong with it?

All e2 teams (except PLi) are working to make repos python 3 compatible also, just take a look at OE-A and you'll see the amount of heavy work for py3

Only recent codes needs another run of PEP8.sh which could be done once a month, if @WanWizard ready for https://github.com/OpenPLi/enigma2/pull/2941 I can send it again and just like .po update we can make repo PEP8 update too.

Forcing PRs for PEP8 will never be 100% equal to what you may think as each distribution/PEP8 tool may do something different. Just run different vetsions of autopep8 tool and you'll see what I mean.

We (Open Vision, Open WebIF and OE-A) are using even PEP8.sh as an automatic tool. It's PLi who's behind this but it's ok as we can run the tool once a month.

**Comment by Huevos** [comment]
> 
> 
> Note that due to some reason there are already few rules broken in the current state of the source files.
> 
>

PEP8 is not rules, it is just a set of guidelines, which are not being followed anyway by these scripts.

**Comment by technic** [comment]
There is a set of guidelines wich were *applied* in the commit I referenced. At the moment almost all scripts do follow them. And other distros apply them automatically on the regular basis.

I can see that there can be friction in changing the code style. But as for now the code style is already changed, and agreed across other enigma2 teams. The linter ci job is a common thing in many open source projects. And to my opinion it is a good strategy to make sure the agreed code style is *maintained*.

**Comment by littlesat** [comment]
Better give attention to new features and really improve something (ci+, offline encoding, making stuff Python 3 ready (not the way it is done now... first agree on how even discuss if it is really required), integrate dab+ rtl-sdr stick...., remove timeshift code completely and change it to make recording and view that recording.... or upgrade to latest oe, or fix and get open HbbTV integrated.
these kind of commit request that were required >15 years ago do not have real added value anymore and somehow ‘proves’ teams are searching for work.... just to create work.... and it also irritates devs over the teams.... 
 
Huevos just responds to.... now we have it again...!!
 
in addition it would also be better now to have less teams so ‘we’ can work on real features.... 
 
Besides PEP I suggest try when you adapt a source file then also adapt that one file (of files) and stop targeting the complete source. So I would recommend for this time stop this project and find read adds for enigma2 and spend the time on those...

**Comment by nautilus7** [comment]
Don't apply any coding rule or quality control (specific coding style, good naming in classes, methods, variables, having no comments, writing every text visible to the user in poor English) and then wonder why developers don't want to spend hours on reading a piece of code, just to fix something broken or add a small new feature...

**Comment by IanSav** [comment]
@nautilus7, Hear, hear!

Well said.  Cleaning up the code and trying to adopt some common style to make it easier to read, easier to change and easier to compare is very long overdue.

The quality of code going into OpenPLi lately has been most disappointing.  We have actually started reverting recent OpenPLi changes merged into OpenVision given the detrimental effects of some changes.

**Comment by Huevos** [comment]
@IanSav, I think you will find he is saying the opposite.

**Comment by IanSav** [comment]
@Huevos, I don't think I am wrong, perhaps @nautilus7 will clarify for us.

Even if I have misunderstood @nautilus7, my comments still stand.  OpenPLi is becoming unstable and potentially unusable as a master upstream source.

**Comment by technic** [comment]
I wouldn't make such pr, if you hadn't agreed on this autopep8 formatter some time ago. So I thought that subset of pep8 lints was decided...

**Comment by Huevos** [comment]
> 
> 
> @Huevos, I don't think I am wrong, perhaps @nautilus7 will clarify for us.
> 
> Even if I have misunderstood @nautilus7, my comments still stand. OpenPLi is becoming unstable and potentially unusable as a master upstream source.

@IanSav, No need for scaremongering. If PLi code is becoming unstable (which I don't believe is true) it has nothing to do with whether the code follows PEP8 guidelines.

PEP8 does not make "unstable" code more stable.

**Comment by nautilus7** [comment]
@IanSav got correctly what I am trying to say...

@Huevos, code that is not easy to read/understand is prone to errors/bugs, thus unstable. I can't tell if Pli is unstable, but following some code styling at least helps readability and finding bugs easier.

**Comment by IanSav** [comment]
No PEP8 does not make unstable code more stable, but it does make it easier to read and see the instability and fix it!

**Comment by persianpros** [comment]
Keep in mind the reason I spent days on sending PRs regarding PEP8 to all enigma2 scene (I hope) was having similar code style so future merges, cherry-picks, diff compares ... would be easier so development/contribution would be easier too.

We should thank all teams that followed this matter first, then we need to run PEP8.sh like once a month to keep things more similar. We do this via Travis but for PLi we could do it by hand like updating po files.

**Comment by technic** [comment]
Well, it's ok to read current enigma2 code. Its a bit annoying because it does not follow usual style. And apparently when writing one has to spend additional mental resources to follow the legacy style.

However some flake8 and few pep8 lints, really rule out bug prone code patterns.

**Comment by Huevos** [comment]
Can you please explain the "by hand" method?

**Comment by persianpros** [comment]
@Huevos 

I mean like running the https://github.com/OpenViX/enigma2/blob/Dev/po/updateallpo-multiOS.sh by hand, once a month is enough to keep the sources PEP8.

Even on Travis which we use is safe but I proposed it before and PLi rejected the automation so only manually will be left.

This won't bring any good to the codes it's just make cherry-picks/merges easier.

**Comment by Huevos** [comment]
I am talking about the Python code, not the .po files.

**Comment by persianpros** [comment]
@Huevos 

I mean run PEP8.sh once a month manually for python files.

**Comment by Huevos** [comment]
We don't have that file available.

**Comment by persianpros** [comment]
@Huevos 

WanWizard should give me the green light for sending the PR, I sent it before but after about two weeks nobody merged it.

**Comment by Huevos** [comment]
> 
> 
> @Huevos
> 
> WanWizard should give me the green light for sending the PR, I sent it before but after about two weeks nobody merged it.

For OpenViX we don't have any script for the .py files like the one for the .po files.

**Comment by persianpros** [comment]
@Huevos 

I was waiting for PLi to merge the PR as it's the upstream so you could benefit too but I can send it for OpenViX.

**Comment by Huevos** [comment]
Are you talking about this: https://github.com/OpenPLi/enigma2/pull/2941/commits/bf86ffb55f199702d45f4ec8721ca89f1c8a0fe5 ?

Is there a guide somewhere to get that up and running?

**Comment by persianpros** [comment]
@Huevos 

https://github.com/OpenViX/enigma2/pull/696

**Comment by TwolDE2** [comment]
Ran this manually on a local clone of my python3 Git with pep8.sh in the base directory.
Some files (I use git workflow) were reverted to python2) and other files like fonts were removed, although .py files were changed OK
I didn‘t look at Travis and by the time I thought about it, I had changed the Ubuntu system
I then found that there is a python3-autopep8, which has to be installed on Ubuntu 20.04
Once installed and the script rerun on a new clone, everything was OK.

Bottom line is that either only install in a directory for python files or treat python2 & python3 branches differently and run manually

**Comment by persianpros** [comment]
I just added some new tools to our repos: https://github.com/OpenVisionE2/enigma2-openvision/commit/ebdf7bbcea951f940671b2499cf984fdda0c9bd6

dos2unix fixes files that came from windows time to time by contributors.

chmod fixes permission changes by contributors and in some .bb recipes we use cp instead of install command with 0644 or 0755 so it's needed.

futurize fixes lots of old python codes to be compatible with both py2 and py3 not just py3, it makes so many sources ready for new versions so no other branch will be needed for them also it makes codes similar to each other just like PEP8 so diffs, compares, cherry-picks and merges would be easier across branches and other repos.

**Comment by WanWizard** [comment]
I am in favour of code standards. Any will do, as long as everyone uses the same :grin: ,

@littlesat ok to merge this?

**Comment by Huevos** [comment]
This pull request is going to stop people contributing to PLi. And it also doesn't show any code is good or logical.

Just makes things more difficult for people that are providing real code with real features and real value.

No idea why you would want this when you can just run the script.

Stifling!

**Comment by technic** [comment]
> This pull request is going to stop people contributing to PLi. And it also doesn't show any code is good or logical.
> 
> Just makes things more difficult for people that are providing real code with real features and real value.
> 
> No idea why you would want this when you can just run the script.

Firstly, I think there is a solution. We can make this CI job not fail, but produce some sort of warning message. Then you can either ask the PR author to run the script, or you can run it yourself and fix the PR, or you can just ignore this warnings for a while.

Secondly, a lot of opensource project adopt code style CI linting strategy, and do not have any problems. Why do you think that enigma2 contributors will have issues?

Upd:

Finally, creating clean commits from the begging is better. When you run the script once in a time this creates noise commits, git blame becomes more polluted.

**Comment by WanWizard** [comment]
> Just makes things more difficult for people that are providing real code with real features and real value.
> No idea why you would want this when you can just run the script.

What is stopping these "people" to run that script before they create the PR, so the code offered adheres to the standards and doesn't fail the CI?

I find this a bit of a non-discussion to be honest, like the famous "spaces vs tabs" war in the PHP community. I prefer tabs, but the majority has chosen for spaces, and most PHP projects don't accept PR's that don't follow the coding standards of that project. 

So be it. You win some, you lose some. You accept. You adapt. You move on.

**Comment by persianpros** [comment]
I hope you like the red sign on your github actions, I'm pretty sure all other teams will revert/skip this change.

**Comment by technic** [comment]
> 
> 
> I hope you like the red sign on your github actions, I'm pretty sure all other teams will revert/skip this change.

I will go after other teams ]:-}

**Comment by WanWizard** [comment]
> I hope you like the red sign on your github actions, I'm pretty sure all other teams will revert/skip this change.

I am very confused now, earlier you wrote "I am waiting for this PR to be merged", and now it is a bad PR? 

Make up your minds people!

**Comment by persianpros** [comment]
@WanWizard 

Where did I wrote such thing?

From the beginning of this PR I was against it as it's the most stupid commit of the year, no offence to anyone but your github actions is fucked now.

**Comment by littlesat** [comment]
Now I'm extreme confused.... but from IanSav I also heard we better can revert it...

@Wanwizard please revert it...

**Comment by persianpros** [comment]
Guys I don't want to consider myself as a pro on PEP8 but I'm pretty sure that I know enough because even OE itself accepted my PEP8 changes and I'm in contact with them for applying my changes on bitbake and core of the OE as well so when I tell you something about PEP8 it's for your own sake.

**Comment by WanWizard** [comment]
Reverted.

**Comment by technic** [comment]
I don't understand enigma2 / OE world. You guys have your own ways :-/
